### ISSUE TITLE
nginx and two other services boot loops

### ISSUE THREAD

**Issue by digitalknk** [spantaleev/matrix-docker-ansible-deploy#966]
The issue is that `matrix-appservice-webhooks`,  `matrix-dimension`, and `matrix-nginx-proxy` end up in a consistent boot loop and none of them ever match up.

I am pretty sure this change ([b89936afba9f922f83dfe771d725518bb1c6b35f](https://github.com/spantaleev/matrix-docker-ansible-deploy/commit/b89936afba9f922f83dfe771d725518bb1c6b35f)) broke my installation, I did a checkout prior to that change ([19cbdce6d22a17c2966b1f9aa5e8ad65c4b355b2](https://github.com/spantaleev/matrix-docker-ansible-deploy/commit/19cbdce6d22a17c2966b1f9aa5e8ad65c4b355b2)) and it works fine.

**Comment by spantaleev** [comment]
Can you provide some logs of all these services trying to start and then crash?

- `matrix-nginx-proxy` requires nothing.
- `matrix-dimension` requires none of the others (just `matrix-postgres.service`)
- `matrix-appservice-webhooks` requires `matrix-nginx-proxy.service` and `matrix-synapse.service`
- `matrix-synapse` requires `matrix-postgres.service`

The first 2 should start somewhat simultaneously, with `matrix-appsrevice-webhooks` following later (because it requires `matrix-nginx-proxy.service` to have been started).

Are you running a setup with Synapse workers enabled? If so, then `matrix-nginx-proxy` may need to wait on `matrix-synapse` as well.

**Comment by enko** [comment]
Logs from nginx:

```
Mar 25 09:46:42 matrix.schumacher.im systemd[1]: Started Matrix nginx-proxy server.
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: 10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf is not a file or does not exist
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: /docker-entrypoint.sh: Configuration complete; ready for start up
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: 2021/03/25 08:46:43 [emerg] 1#1: host not found in upstream "matrix-appservice-webhooks" in /etc/nginx/conf.d/matrix-domain.conf:73
Mar 25 09:46:43 matrix.schumacher.im matrix-nginx-proxy[11140]: nginx: [emerg] host not found in upstream "matrix-appservice-webhooks" in /etc/nginx/conf.d/matrix-domain.conf:73
Mar 25 09:46:43 matrix.schumacher.im systemd[1]: matrix-nginx-proxy.service: Main process exited, code=exited, status=1/FAILURE
```

Logs from webhooks:

```
Mar 25 09:50:55 matrix.schumacher.im matrix-appservice-webhooks[21484]: matrix-appservice-webhooks
Mar 25 09:50:55 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Main process exited, code=exited, status=137/n/a
Mar 25 09:50:55 matrix.schumacher.im systemd[1]: Stopped Matrix Appservice webhooks bridge.
Mar 25 09:50:55 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Unit entered failed state.
Mar 25 09:50:55 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Failed with result 'exit-code'.
Mar 25 09:50:56 matrix.schumacher.im systemd[1]: Starting Matrix Appservice webhooks bridge...
Mar 25 09:51:01 matrix.schumacher.im systemd[1]: Started Matrix Appservice webhooks bridge.
Mar 25 09:51:02 matrix.schumacher.im matrix-appservice-webhooks[22144]: Unable to load crypto module: crypto will be disabled: Error: global.Olm is not defined
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.191 +00:00 info [index] Preparing database...
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.196 +00:00 info [WebhookStore] Running migrations
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: [INFO] No migrations to run
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: [INFO] Done
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.369 +00:00 warn [WebhookStore] Changing database path to be data/production.db to ensure data is persisted
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.412 +00:00 info [index] Preparing bridge...
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.413 +00:00 info [WebhookBridge] Constructing bridge
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.415 +00:00 info [WebhookBridge] Starting bridge
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.531 +00:00 info [WebhookBridge] Request to bridge room !bFxuhjTBJVBLiawCur:schumacher.im
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.534 +00:00 info [WebhookBridge] Request to bridge room !twifNuOvtgJGmSVFEI:schumacher.im
Mar 25 09:51:03 matrix.schumacher.im matrix-appservice-webhooks[22144]: Mar-25-2021 08:51:03.535 +00:00 info [WebhookBridge] Request to bridge room !kdByTLdjdkPEMZnLDT:schumacher.im
Mar 25 09:51:27 matrix.schumacher.im systemd[1]: Stopping Matrix Appservice webhooks bridge...
Mar 25 09:51:28 matrix.schumacher.im matrix-appservice-webhooks[22743]: matrix-appservice-webhooks
Mar 25 09:51:28 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Main process exited, code=exited, status=137/n/a
Mar 25 09:51:28 matrix.schumacher.im systemd[1]: Stopped Matrix Appservice webhooks bridge.
Mar 25 09:51:28 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Unit entered failed state.
Mar 25 09:51:28 matrix.schumacher.im systemd[1]: matrix-appservice-webhooks.service: Failed with result 'exit-code'.
Mar 25 09:51:28 matrix.schumacher.im systemd[1]: Starting Matrix Appservice webhooks bridge...
Mar 25 09:51:33 matrix.schumacher.im systemd[1]: Started Matrix Appservice webhooks bridge.
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Unable to load crypto module: crypto will be disabled: Error: global.Olm is not defined
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.699 +00:00 info [index] Preparing database...
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.701 +00:00 info [WebhookStore] Running migrations
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: [INFO] No migrations to run
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: [INFO] Done
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.848 +00:00 warn [WebhookStore] Changing database path to be data/production.db to ensure data is persisted
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.915 +00:00 info [index] Preparing bridge...
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.917 +00:00 info [WebhookBridge] Constructing bridge
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.919 +00:00 info [WebhookBridge] Starting bridge
Mar 25 09:51:35 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.998 +00:00 info [WebhookBridge] Request to bridge room !bFxuhjTBJVBLiawCur:schumacher.im
Mar 25 09:51:36 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:35.999 +00:00 info [WebhookBridge] Request to bridge room !twifNuOvtgJGmSVFEI:schumacher.im
Mar 25 09:51:36 matrix.schumacher.im matrix-appservice-webhooks[23407]: Mar-25-2021 08:51:36.000 +00:00 info [WebhookBridge] Request to bridge room !kdByTLdjdkPEMZnLDT:schumacher.im
```

Logs from dimension:

```
Mar 25 09:51:04 matrix.schumacher.im systemd[1]: Starting Matrix Dimension...
Mar 25 09:51:04 matrix.schumacher.im systemd[1]: Started Matrix Dimension.
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: version Error: [git-rev-sync] no git repository found
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at _getGitDirectory (/home/node/matrix-dimension/node_modules/git-rev-sync/index.js:47:11)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at _getGitDirectory (/home/node/matrix-dimension/node_modules/git-rev-sync/index.js:75:10)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at branch (/home/node/matrix-dimension/node_modules/git-rev-sync/index.js:79:16)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at long (/home/node/matrix-dimension/node_modules/git-rev-sync/index.js:92:11)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Object.short (/home/node/matrix-dimension/node_modules/git-rev-sync/index.js:118:10)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Object.<anonymous> (/home/node/matrix-dimension/build/app/version.js:14:19)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Module._compile (internal/modules/cjs/loader.js:1158:30)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Module.load (internal/modules/cjs/loader.js:1002:32)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Function.Module._load (internal/modules/cjs/loader.js:901:14)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Module.require (internal/modules/cjs/loader.js:1044:19)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at require (internal/modules/cjs/helpers.js:77:18)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Object.<anonymous> (/home/node/matrix-dimension/build/app/index.js:15:19)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Module._compile (internal/modules/cjs/loader.js:1158:30)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Module.load (internal/modules/cjs/loader.js:1002:32)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Function.Module._load (internal/modules/cjs/loader.js:901:14)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]:     at internal/main/run_main_module.js:18:47
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:06.750 +00:00 info [index] Starting dimension v1.0.0
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:06.751 +00:00 info [DimensionStore] Updating schema...
Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:06.813 +00:00 verbose [DimensionStore [SQL]] Executing (default): CREATE TABLE IF NOT EXISTS "SequelizeMeta" ("name" VARCHAR(255) Mar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:06.827 +00:00 verbose [DimensionStore [SQL]] Executing (default): SELECT i.relname AS name, ix.indisprimary AS primary, ix.indisunMar 25 09:51:06 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:06.837 +00:00 verbose [DimensionStore [SQL]] Executing (default): SELECT "name" FROM "SequelizeMeta" AS "SequelizeMeta" ORDER BY "
[...removed all registering routes log entry...]
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:07.099 +00:00 info [Webserver] API and UI listening on 0.0.0.0:8184
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:07.099 +00:00 info [BotSdk-MatrixLiteClient (REQ-1)] GET http://matrix-nginx-proxy:12080/_matrix/client/r0/account/whoami
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: Mar-25-2021 08:51:07.110 +00:00 error [BotSdk-MatrixLiteClient (REQ-1)] getaddrinfo ENOTFOUND matrix-nginx-proxy
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: Error: getaddrinfo ENOTFOUND matrix-nginx-proxy
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:     at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:64:26)
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: Error: getaddrinfo ENOTFOUND matrix-nginx-proxy
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:     at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:64:26) {
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:   errno: 'ENOTFOUND',
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:   code: 'ENOTFOUND',
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:   syscall: 'getaddrinfo',
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]:   hostname: 'matrix-nginx-proxy'
Mar 25 09:51:07 matrix.schumacher.im matrix-dimension[22563]: }
```

**Comment by spantaleev** [comment]
Looks like the way `matrix-appservice-webhooks` injects its configuration into `matrix-nginx-proxy` is a little fragile: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/59ac8a59210814ecd2834b4603ae61cfc91ff413/roles/matrix-bridge-appservice-webhooks/tasks/init.yml#L50-L51

`proxy_pass` with a URL like this makes nginx try to resolve the hostname at nginx start time. If the service is unavailable, nginx won't start. Also, it only resolves it once. If that services gets restarted later on and appears on a new IP address, nginx will still try to contact it on the old one. We should probably redo this.

Normally, we use something like this: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/59ac8a59210814ecd2834b4603ae61cfc91ff413/roles/matrix-nginx-proxy/templates/nginx/conf.d/matrix-client-element.conf.j2#L16-L18

.. which makes DNS resolution happen later on, on-demand and once every 5 seconds.

-----

I guess we need to use something like this in `roles/matrix-bridge-appservice-webhooks/tasks/init.yml`:

```diff
  resolver 127.0.0.11 valid=5s;
- proxy_pass {{ matrix_appservice_webhooks_appservice_url }}:{{ matrix_appservice_webhooks_matrix_port }}/;
+ set $backend "matrix-appservice-webooks:{{ matrix_appservice_webhooks_matrix_port }}";
+ proxy_pass http://$backend/;
```

I don't know if we can put a trailing `/` in `proxy_pass` when it uses a variable though. If anyone is using this, you can try a few things like that and let me know what works (or submit a PR).

----------

Alternatively, we can make `matrix-nginx-proxy.service` depend on `matrix-appservice-webhooks` (if `matrix_appservice_webhooks_enabled: true`) by modifying `matrix_nginx_proxy_systemd_required_services_list` via `group_vars/matrix_servers`, but that's not ideal.

It would fix the ordering problem, but nginx would still resolve the address of `matrix-appservice-webhooks` just once, on startup. If the service gets a new IP address, nginx would not pick up on it.

It's better if we go with the other solution.

**Comment by skepticalwaves** [comment]
I got bitten by this as well and had to revert to the earlier commit (thanks for providing it!)

**Comment by throwawayay** [comment]
Same here, reverted with: `git checkout 19cbdce6d22a17c2966b1f9aa5e8ad65c4b355b2` temporarily

**Comment by skepticalwaves** [comment]
@spantaleev 
> I don't know if we can put a trailing / in proxy_pass when it uses a variable though. If anyone is using this, you can try a few things like that and let me know what works (or submit a PR).

I *think* this is how the variable-based proxy_pass needs to be configured:
https://trac.nginx.org/nginx/ticket/1067

**Comment by MarcProe** [comment]
same problem here, I removed the webhooks for now. :-(

**Comment by MarcProe** [comment]
I did a rather straightforward change that, however, does **_not_** work:
```diff
-          proxy_pass {{ matrix_appservice_webhooks_appservice_url }}:{{ matrix_appservice_webhooks_matrix_port }}/;
+          set $backend "{{ matrix_appservice_webhooks_appservice_url }}:{{ matrix_appservice_webhooks_matrix_port }}";
+          proxy_pass $backend;
```

It gives a ```404```-error: ```Cannot POST /```

Removing the trailing ```/```:
```diff
-          proxy_pass {{ matrix_appservice_webhooks_appservice_url }}:{{ matrix_appservice_webhooks_matrix_port }};
+          set $backend "{{ matrix_appservice_webhooks_appservice_url }}:{{ matrix_appservice_webhooks_matrix_port }}";
+          proxy_pass $backend;
```
...gives ```404``` with ```Cannot POST /appservice-webhooks/api/v1/matrix/hook/hErEiStHeHoOkId```

**Comment by MarcProe** [comment]
> I am pretty sure this change ([b89936afba9f922f83dfe771d725518bb1c6b35f](https://github.com/spantaleev/matrix-docker-ansible-deploy/commit/b89936afba9f922f83dfe771d725518bb1c6b35f)) broke my installation, I did a checkout prior to that change ([19cbdce6d22a17c2966b1f9aa5e8ad65c4b355b2](https://github.com/spantaleev/matrix-docker-ansible-deploy/commit/19cbdce6d22a17c2966b1f9aa5e8ad65c4b355b2)) and it works fine.

Reverting the breaking change restores functionality, which might be the quickest fix for anybody using the module until there's a real solution.
```diff
diff --git a/group_vars/matrix_servers b/group_vars/matrix_servers
index e1d1239c..abaab696 100755
--- a/group_vars/matrix_servers
+++ b/group_vars/matrix_servers
@@ -117,8 +117,6 @@ matrix_appservice_webhooks_systemd_required_services_list: |
     ['docker.service']
     +
     (['matrix-synapse.service'] if matrix_synapse_enabled else [])
-    +
-    (['matrix-nginx-proxy.service'] if matrix_nginx_proxy_enabled else [])
   }}

 ######################################################################
```

**Comment by aaronraimist** [comment]
@MarcProe are you using the built in nginx or did you disable it?

**Comment by MarcProe** [comment]
I'm using the default role, ```matrix-ngninx-proxy```.

**Comment by MarcProe** [comment]
Tried a fix, please test #987 .

**Comment by spantaleev** [comment]
Supposedly fixed by #987. If not, please reopen.
